src/app/AttributePathExpandIterator.h

```diff
@@ -125,5 +148,59 @@ class AttributePathExpandIterator
     bool IsValidAttributeId(AttributeId attributeId);
 };
 
+/// PeekAttributePathExpandIterator is an AttributePathExpandIterator wrapper that rolls back the Next()
+/// call whenever a new `MarkCompleted()` method is not called (until Next() returns false). This is useful
+/// to allow pausing iteration in cases the next path was not ready to be used yet and iteration needs to
+/// continue later.
+///
+/// Example use cases:
+///
+/// - Iterate over all attributes and process one-by-one, however when the iteration fails, resume at
+///   the last failure point:
+///
+///      PeekAttributePathExpandIterator iterator(....);
+///      ConcreteAttributePath path;
+///
+///      while (iterator.Next(path)) {
+///         if (!CanProcess(path)) {
+///             // iterator state IS PRESERVED so that Next() will return the SAME path on the next call.
+///             return CHIP_ERROR_TRY_AGAIN_LATER;
+///         }
+///      }
+///      // this will make Next() NOT return the previous value.
+///      iterator.MarkCompleted();
+///
+/// -  Grab what the next output path would be WITHOUT advancing a state;
+///
+///      {
+///        PeekAttributePathExpandIterator iterator(...., state);
+///        if (iterator.Next(...)) { ... }
+///      }
+///      // state here is ROLLED BACK (i.e. next calls the same value)
```

Comment:
```suggestion
///      // state here is ROLLED BACK (i.e. initializing a new iterator with it will start at the same place as the previous iteration attempt).
```

or something?